### ISSUE TITLE
[rllib] Mismatch between evaluation and training metrics when using evaluation_parallel_to_training

### ISSUE THREAD

**Issue by loicsacre** [ray-project/ray#17924]
### What is the problem?

When using the `evaluation_parallel_to_training` agent parameter, an evaluation is launched at the same time as a training iteration. See https://github.com/ray-project/ray/blob/7e109a3266016a35501ddd897f279576fe7bfc3d/rllib/agents/trainer_template.py#L203-L207

Therefore, for instance, at the first train call when getting back the results in Tune callbacks, the evaluation results correspond to an evaluation of a non-trained network (iteration 0) while the training results (the loss, etc) correspond to the first training iteration. It is thus difficult to log correctly to Wandb for instance as the eval results corresponds to the iteration 0 while the training ones to the iteration 1. A potential solution would be to cache the eval results and get them back in the following train call.

*Ray version and other system information (Python version, TensorFlow version, OS):*

Ray: 1.5.0
Python 3.8
Pytorch 1.9
Linux/MacOs

### Reproduction (REQUIRED)
When using the `evaluation_parallel_to_training` agent parameter.

**Comment by gjoliver** [comment]
I can take this one.

**Comment by sven1977** [comment]
Hey @loicsacre , thanks for filing this. :) I'm not sure I understand the caching suggestion. It's correct that the evaluation in case you do parallel eval+training is always done on the "previous" version of the trainer (model). So if we cached the eval results and only published them 1 iteration later, wouldn't that mean that the eval results are then 2 iterations behind the train results? Do you mean to cache the train results?

**Comment by loicsacre** [comment]
Hey @sven1977, yes sorry I was mistaken. Indeed, it would be better to cache the train results of course. I don't know if it is the best solution but the idea is to be able to log the eval results corresponding to the right training iteration as you have well understood.
